Support img tags in HTML label text

nyalldawson opened the pull request:
Allows use of img tags in HTML label content. The following logic is applied:

- Image path is set via the src="xxx" attribute. Local, HTTP, and base64 encoded paths are permitted
- Any image format readable by QGIS can be used
- Image sizes can be specified via the width="##" and height="##" attributes. If width or height is not specified it will automatically be calculated from the original image size
- If width or height are specified, they are considered to be in POINTS
- The css width/height settings are NOT respected (this is a Qt limitation)
- Images are not supported for curved text labels
- Images are placed inline only, floating images are not supported

Sponsored by City of Freiburg im Breisgau

elpaso commented on src/core/textrenderer/qgstextblockformat.h:
```diff
@@ -0,0 +1,131 @@
+  private:
+
+    bool mHasHorizontalAlignSet = false;
+    Qgis::TextHorizontalAlignment mHorizontalAlign = Qgis::TextHorizontalAlignment::Left;
```
Is it RTL relevant here? Shouldn't the default be adapted in that case?

nyalldawson committed:
Allow HTML data urls for QgsImageCache

Adds support for base64 encoded image decoding when the path
is a HTML data URL (in addition to the existing "base64:..."
format support)

Allows use of eg "data:image/jpeg;base64,XXXXXXXX" formats
for image paths, so that the image cache can correctly handle
embedded image paths from HTML/CSS content

nyalldawson committed:
Allow QgsTextFragment to represent an inline image object

Modifies the QgsTextDocument API to extract images from HTML
content, and store in the associated
QgsTextFragment/QgsTextCharacterFormat objects

nyalldawson committed:
[feature] Support img tags in HTML label text

Allows use of img tags in HTML label content. The following logic
is applied:

- Image path is set via the src="xxx" attribute. Local, HTTP, and
base64 encoded paths are permitted
- Any image format readable by QGIS can be used
- Image sizes can be specified via the width="##" and height="##"
attributes. If width or height is not specified it will automatically
be calculated from the original image size
- If width or height are specified, they are considered to be
in POINTS
- The css width/height settings are NOT respected (this is a Qt
limitation)
- Images are not supported for curved text labels
- Images are placed inline only, floating images are not supported

Sponsored by City of Freiburg im Breisgau

nyalldawson committed:
Test masks
